rust/src/applayertemplate/template.rs

```diff
@@ -96,7 +96,7 @@ impl TemplateState {
         let mut index = 0;
         for i in 0..len {
             let tx = &self.transactions[i];
-            if tx.tx_id == tx_id + 1 {
+            if tx.tx_id == tx_id {
```

Comment:
I think this needs a closer look.

A TX should never have an ID of 0, the first one created gets an ID of 1.  However, I believe the engine requests transactions at a 0 offset, so you have to add 1 to find the correct TX.  This is how it used to be at least, and is why the template is like this.  Other parsers like DNS and MQTT also do what the template does and appear to be correct.

The reason the first ID is 1 and not 0 is because the engine considered 0 to mean no TX.

So if this is to change in template, all the parsers that follow this pattern probably need to be visited as well for correctness.

Reply:
Yeah, just confirmed The engine will call the parsers state for a TX ID of 0, but inside the parser, this is ID 1.

Reply:
Sorry about the extra work, I'll leave the template change out 🙇🏽

Reply:
The interesting thing about this is that for pgsql, the engine was checking for the correct id, this is what led to me checking and finding the bug with the SV tests for pgsql...

Reply:
I'd like to look into it further.  But I think if you look closer, you'll see that pgsql still creates its first tx_id as 1, not 0.  Then the engine comes and requests 0 but will never find one.

Its been a while since I've looked at this closely (years actually), but it probably deserves some documentation.

Reply:
We have the place for that, but I'd have to learn more to document it well: https://docs.suricata.io/en/latest/devguide/extending/app-layer/transactions.html
I'll debug a big with pgsql, see if I manage to extract the needed info and give it a try.

Reply:
I _think_... that when the engine calls the app-parser logging function, it indeed calls with tx_id starting from 0. But when the alert logging function is called, it has access to the right tx_id, as that comes from the packetalert, and this is what was messing with pgsql alert metadata 🤔

Reply:
Added a bit more context on what I learned here: https://github.com/OISF/suricata/pull/11008#issuecomment-2089189407

Maybe part of that could be incorporated into the documentation?